rust_snuba/src/strategies/clickhouse/writer_v2.rs

```diff
@@ -0,0 +1,215 @@
+            let write_finish = SystemTime::now();
+
+            if let Ok(elapsed) = write_finish.duration_since(write_start) {
+                timer!("insertions.batch_write_ms", elapsed);
+            }
+            counter!("insertions.batch_write_msgs", batch_len as i64);
+            empty_batch.record_message_latency();
+
+            Ok(empty_message.replace(empty_batch))
+        })
+    }
+}
+
+pub struct ClickhouseWriterStep<N> {
+    inner: RunTaskInThreads<BytesInsertBatch<RowData>, BytesInsertBatch<()>, anyhow::Error, N>,
+}
+
+impl<N> ClickhouseWriterStep<N>
+where
+    N: ProcessingStrategy<BytesInsertBatch<()>> + 'static,
+{
+    pub fn new(
+        next_step: N,
+        cluster_config: ClickhouseConfig,
+        table: String,
+        skip_write: bool,
+        concurrency: &ConcurrencyConfig,
+    ) -> Self {
+        let hostname = cluster_config.host;
+        let http_port = cluster_config.http_port;
+        let database = cluster_config.database;
+
+        let inner = RunTaskInThreads::new(
+            next_step,
+            clickhouse_task_runner(
+                Arc::new(ClickhouseClient::new(
+                    &hostname, http_port, &table, &database,
+                )),
+                skip_write,
+            ),
+            concurrency,
+            Some("clickhouse"),
+        );
+
+        ClickhouseWriterStep { inner }
+    }
+}
+
+impl<N> ProcessingStrategy<BytesInsertBatch<RowData>> for ClickhouseWriterStep<N>
+where
+    N: ProcessingStrategy<BytesInsertBatch<()>>,
+{
+    fn poll(&mut self) -> Result<Option<CommitRequest>, StrategyError> {
+        self.inner.poll()
+    }
+
+    fn submit(
+        &mut self,
+        message: Message<BytesInsertBatch<RowData>>,
+    ) -> Result<(), SubmitError<BytesInsertBatch<RowData>>> {
+        self.inner.submit(message)
+    }
+
+    fn terminate(&mut self) {
+        self.inner.terminate();
+    }
+
+    fn join(&mut self, timeout: Option<Duration>) -> Result<Option<CommitRequest>, StrategyError> {
+        self.inner.join(timeout)
+    }
+}
+
+#[derive(Clone)]
+pub struct ClickhouseClient {
+    client: Client,
+    headers: HeaderMap<HeaderValue>,
+    url: String,
+    query: String,
+}
+
+impl ClickhouseClient {
+    pub fn new(hostname: &str, http_port: u16, table: &str, database: &str) -> ClickhouseClient {
+        let mut headers = HeaderMap::with_capacity(3);
+        headers.insert(CONNECTION, HeaderValue::from_static("keep-alive"));
+        headers.insert(ACCEPT_ENCODING, HeaderValue::from_static("gzip,deflate"));
+        headers.insert(
+            "X-ClickHouse-Database",
+            HeaderValue::from_str(database).unwrap(),
+        );
+
+        let query_params = "load_balancing=in_order&insert_distributed_sync=1".to_string();
+        let url = format!("http://{hostname}:{http_port}?{query_params}");
+        let query = format!("INSERT INTO {table} FORMAT JSONEachRow");
+
+        ClickhouseClient {
+            client: Client::new(),
```

Comment:
Does it make sense to add a request timeout `.timeout(Duration::from_secs(x))` and make it configurable via `ClickhouseConfig`?